File: Sistema de Calidad/app/src/main/java/com/example/sistemadecalidad/utils/TimeUtils.kt
package com.example.sistemadecalidad.utils

import java.text.SimpleDateFormat
import java.util.*

/**
 * Utilidades para manejo consistente de tiempo en zona horaria de Perú
 * Garantiza que toda la aplicación use la misma zona horaria (America/Lima)
 */
object TimeUtils {
    
    private const val PERU_TIMEZONE = "America/Lima"
    private val peruTimeZone = TimeZone.getTimeZone(PERU_TIMEZONE)
    
    /**
     * Obtiene la fecha actual en zona horaria de Perú
     */
    fun getCurrentPeruDate(): Date {
        val calendar = Calendar.getInstance(peruTimeZone)
        return calendar.time
    }
    
    /**
     * Formatea una fecha para mostrar en formato de fecha legible en español
     * Ejemplo: "lunes, 15 de enero"
     */
    fun formatDateForDisplay(date: Date = getCurrentPeruDate()): String {
        val formatter = SimpleDateFormat("EEEE, dd 'de' MMMM", Locale("es", "PE"))
        formatter.timeZone = peruTimeZone
        return formatter.format(date)
    }
    
    /**
     * Formatea una fecha para mostrar la hora en formato HH:mm:ss
     * Ejemplo: "14:30:25"
     */
    fun formatTimeForDisplay(date: Date = getCurrentPeruDate()): String {
        val formatter = SimpleDateFormat("HH:mm:ss", Locale("es", "PE"))
        formatter.timeZone = peruTimeZone
        return formatter.format(date)
    }
    
    /**
     * Formatea una fecha para envío al backend en formato YYYY-MM-DD
     * Ejemplo: "2024-01-15"
     */
    fun formatDateForBackend(date: Date = getCurrentPeruDate()): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd", Locale("es", "PE"))
        formatter.timeZone = peruTimeZone
        return formatter.format(date)
    }
    
    /**
     * Formatea una hora para envío al backend en formato HH:mm:ss
     * Ejemplo: "14:30:25"
     */
    fun formatTimeForBackend(date: Date = getCurrentPeruDate()): String {
        val formatter = SimpleDateFormat("HH:mm:ss", Locale("es", "PE"))
        formatter.timeZone = peruTimeZone
        return formatter.format(date)
    }
    
    /**
     * Obtiene un timestamp ISO en zona horaria de Perú
     * Ejemplo: "2024-01-15T14:30:25.123-05:00"
     */
    fun getPeruTimestamp(date: Date = getCurrentPeruDate()): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale("es", "PE"))
        formatter.timeZone = peruTimeZone
        return formatter.format(date)
    }
    
    /**
     * Convierte una fecha UTC a zona horaria de Perú
     */
    fun convertUtcToPeruTime(utcDate: Date): Date {
        val utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        utcCalendar.time = utcDate
        
        val peruCalendar = Calendar.getInstance(peruTimeZone)
        peruCalendar.timeInMillis = utcCalendar.timeInMillis
        
        return peruCalendar.time
    }
    
    /**
     * Obtiene la zona horaria de Perú
     */
    fun getPeruTimeZone(): TimeZone = peruTimeZone
    
    /**
     * Verifica si una fecha está en el mismo día (en zona horaria de Perú)
     */
    fun isSameDay(date1: Date, date2: Date): Boolean {
        val cal1 = Calendar.getInstance(peruTimeZone)
        cal1.time = date1
        
        val cal2 = Calendar.getInstance(peruTimeZone)
        cal2.time = date2
        
        return cal1.get(Calendar.YEAR) == cal2.get(Calendar.YEAR) &&
               cal1.get(Calendar.DAY_OF_YEAR) == cal2.get(Calendar.DAY_OF_YEAR)
    }
}

File: Backend/utils/timeUtils.js
/**
 * Utilidades para manejo consistente de tiempo en zona horaria de Perú
 * Garantiza que todo el backend use la misma zona horaria (America/Lima)
 */

const PERU_TIMEZONE = 'America/Lima';

/**
 * Obtiene la fecha actual en zona horaria de Perú
 * @returns {Date} Fecha actual en zona horaria de Perú
 */
function getCurrentPeruDate() {
    const now = new Date();
    return new Date(now.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
}

/**
 * Formatea una fecha para la base de datos en formato YYYY-MM-DD
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {string} Fecha en formato YYYY-MM-DD
 */
function formatDateForDB(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    return peruDate.toISOString().split('T')[0];
}

/**
 * Formatea una hora para la base de datos en formato HH:MM:SS
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {string} Hora en formato HH:MM:SS
 */
function formatTimeForDB(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    return peruDate.toLocaleTimeString('es-PE', { 
        hour: '2-digit', 
        minute: '2-digit', 
        second: '2-digit',
        hour12: false,
        timeZone: PERU_TIMEZONE
    });
}

/**
 * Formatea una fecha para mostrar en formato legible en español
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {string} Fecha en formato legible
 */
function formatDateForDisplay(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    return peruDate.toLocaleDateString('es-ES', {
        weekday: 'long',
        year: 'numeric',
        month: 'long',
        day: 'numeric',
        timeZone: PERU_TIMEZONE
    });
}

/**
 * Formatea una hora para mostrar en formato 12 horas
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {string} Hora en formato 12 horas
 */
function formatTimeFor12Hour(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    return peruDate.toLocaleTimeString('es-ES', {
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: true,
        timeZone: PERU_TIMEZONE
    });
}

/**
 * Obtiene un timestamp ISO en zona horaria de Perú
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {string} Timestamp ISO
 */
function getPeruTimestamp(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    return peruDate.toISOString();
}

/**
 * Obtiene información completa de tiempo para respuestas de API
 * @param {Date} date - Fecha a formatear (opcional, usa fecha actual si no se proporciona)
 * @returns {Object} Objeto con información completa de tiempo
 */
function getTimeInfo(date = getCurrentPeruDate()) {
    const peruDate = new Date(date.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    
    return {
        timestamp: getPeruTimestamp(peruDate),
        fecha: formatDateForDB(peruDate),
        hora: formatTimeForDB(peruDate),
        fecha_completa: formatDateForDisplay(peruDate),
        hora_12h: formatTimeFor12Hour(peruDate),
        dia_semana: peruDate.toLocaleDateString('es-ES', { weekday: 'long', timeZone: PERU_TIMEZONE }),
        mes: peruDate.toLocaleDateString('es-ES', { month: 'long', timeZone: PERU_TIMEZONE }),
        año: peruDate.getFullYear(),
        dia_numero: peruDate.getDate(),
        mes_numero: peruDate.getMonth() + 1,
        dia_semana_numero: peruDate.getDay(),
        timezone: PERU_TIMEZONE,
        timezone_offset: -5, // UTC-5 para Perú
        unix_timestamp: Math.floor(peruDate.getTime() / 1000),
        milliseconds: peruDate.getTime()
    };
}

/**
 * Verifica si dos fechas están en el mismo día (en zona horaria de Perú)
 * @param {Date} date1 - Primera fecha
 * @param {Date} date2 - Segunda fecha
 * @returns {boolean} True si están en el mismo día
 */
function isSameDay(date1, date2) {
    const peru1 = new Date(date1.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    const peru2 = new Date(date2.toLocaleString('en-US', { timeZone: PERU_TIMEZONE }));
    
    return peru1.toDateString() === peru2.toDateString();
}

module.exports = {
    PERU_TIMEZONE,
    getCurrentPeruDate,
    formatDateForDB,
    formatTimeForDB,
    formatDateForDisplay,
    formatTimeFor12Hour,
    getPeruTimestamp,
    getTimeInfo,
    isSameDay
};